examples/step-69/step-69.cc

```diff
@@ -0,0 +1,2893 @@
+    // This, unfortunately, does not play nicely with a global distributed
+    // index range because a sparsity pattern with CSR cannot contain
+    // "holes" in the index range. The distributed matrices offered by
+    // deal.II avoid this by translating from a global index range into a
+    // contiguous local index range. But this is the precisely the type of
+    // index manipulation we want to avoid in our assembly loops.
+    //
+    // The Utilities::MPI::Partitioner already implements the translation
+    // from a global index range to a contiguous local (per MPI rank) index
+    // range: we don't have to reinvent the wheel. We just need to use that
+    // translation capability (once and only once) in order to create a
+    // "local" sparsity pattern for the contiguous index range
+    // $[0,$<code>n_locally_relevant</code>$)$. That capability can be
+    // invoked by Utilities::MPI::Partitioner::global_to_local() function.
+    // Once the sparsity pattern is created using local indices, all that
+    // is left to do is to ensure that (when implementing our scatter and
+    // gather auxiliary functions) we always access elements of a
+    // distributed vector by a call to
+    // LinearAlgebra::distributed::Vector::local_element(). This way we
+    // avoid index translations altogether and operate exclusively with
+    // local indices.
+
+    {
+      TimerOutput::Scope t(
+        computing_timer,
+        "offline_data - create sparsity pattern and set up matrices");
+
+      // We have to create the "local" sparsity pattern by hand. We
+      // therefore loop over all locally owned and ghosted cells (see @ref
+      // GlossArtificialCell) and extract the (global)
+      // <code>dof_indices</code> associated to the cell DOFs and renumber
+      // them using <code>partitioner->global_to_local(index)</code>.
+      //
+      // @note In the case of a locally owned dof, such renumbering consist
+      // of applying a shift (i.e. we subtract an offset) such that now they
+      // will become a number in the integer interval
+      // $[0,$<code>n_locally_owned</code>$)$. However, in the case of a
+      // ghosted dof (i.e. not locally owned) the situation is quite
+      // different, since the global indices associated to ghosted DOFs will
+      // not be (in general) a contiguous set of integers.
+
+      DynamicSparsityPattern dsp(n_locally_relevant, n_locally_relevant);
+
+      std::vector<types::global_dof_index> dof_indices(dofs_per_cell);
+
+      for (const auto &cell : dof_handler.active_cell_iterators())
+        {
+          if (cell->is_artificial())
+            continue;
+
+          cell->get_dof_indices(dof_indices);
+          std::transform(dof_indices.begin(),
+                         dof_indices.end(),
+                         dof_indices.begin(),
+                         [&](auto index) {
+                           return partitioner->global_to_local(index);
```

Comment:
@kronbichler *ping*

For generic containers it would basically precisely be this `std::transform` call. For `IndexSet` we can probably perform a more efficient manipulation. So probably yes, we could provide such a function but it won't be terribly urgent :-)

Reply:
I agree with the issue.